datasource/ethereum/src/block_ingestor.rs

```diff
@@ -0,0 +1,178 @@
+                    missing_block_hashes,
+                    move |missing_block_hashes| -> Box<Future<Item = _, Error = _> + Send> {
+                        if missing_block_hashes.is_empty() {
+                            // If no blocks were missing, then the block head pointer was updated
+                            // successfully, and this poll has completed.
+                            Box::new(future::ok(future::Loop::Break(())))
+                        } else {
+                            // Some blocks are missing: load them, ingest them, and repeat.
+                            let missing_blocks = self.get_blocks(&missing_block_hashes);
+                            Box::new(self.ingest_blocks(missing_blocks).map(
+                                |missing_block_hashes| future::Loop::Continue(missing_block_hashes),
+                            ))
+                        }
+                    },
+                )
+            })
+    }
+
+    fn get_latest_block(&self) -> impl Future<Item = Block<Transaction>, Error = Error> {
+        let web3 = Web3::new(self.web3_transport.clone());
+        web3.eth()
+            .block_with_txs(BlockNumber::Latest.into())
+            .map_err(|e| format_err!("could not get latest block from Ethereum: {}", e))
+    }
+
+    /// Put some blocks into the block store (if they are not there already), and try to update the
+    /// head block pointer. If missing blocks prevent such an update, return a Vec with at least
+    /// one of the missing blocks' hashes.
+    fn ingest_blocks<'a, B: Stream<Item = Block<Transaction>, Error = Error> + Send + 'a>(
+        &'a self,
+        blocks: B,
+    ) -> impl Future<Item = Vec<H256>, Error = Error> + Send + 'a {
+        self.store
+            .lock()
+            .unwrap()
+            .upsert_blocks(&self.network_name, blocks)
+            .and_then(move |()| {
+                self.store
+                    .lock()
+                    .unwrap()
+                    .attempt_head_update(&self.network_name, self.ancestor_count)
+            })
+    }
+
+    /// Requests the specified blocks via web3, returning them in a stream (potentially out of
+    /// order).
+    fn get_blocks<'a>(
+        &'a self,
+        block_hashes: &[H256],
+    ) -> Box<Stream<Item = Block<Transaction>, Error = Error> + Send + 'a> {
+        // Don't bother with a batch request if nothing to request
+        if block_hashes.is_empty() {
+            return Box::new(stream::empty());
+        }
+
+        let web3 = Web3::new(Batch::new(self.web3_transport.clone()));
+
+        // Add requests to batch
+        let block_futures = block_hashes
+            .into_iter()
+            .map(|block_hash| {
+                web3.eth()
+                    .block_with_txs(BlockId::from(*block_hash))
+                    .map_err(|e| format_err!("could not get block from Ethereum: {}", e))
+            })
+            .collect::<Vec<_>>();
```

Comment:
This `collect` can just be removed I think.

Reply:
Ah, weird API, if that's the case then could add a note explaining that need for `collect`?